sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala

```diff
@@ -0,0 +1,416 @@
+    *
+    * For that, they are going through following steps:
+    *
+    * - Bob is generating first protocol message a_Bob and sends it to Alice
+    * - Alice forms a hint which contain Bob's commitment "a_Bob", and puts the hint into a hints bag
+    * - She proves the statement using the bag, getting the partial protocol transcript
+    * (a_Alice, e, z_Alice) as a result and sends "a_Alice" and "z_Alice" to Bob.
+    * Please note that "e" is got from both a_Alice and a_Bob.
+    *
+    * - Bob now also knows a_Alice, so can generate the same "e" as Alice. Thus Bob is generating valid
+    * proof ((a_Alice, a_Bob), e, (z_Alice, z_Bob)).
+    */
+  property("distributed AND (2 out of 2)") {
+    val proverA = new ErgoLikeTestProvingInterpreter
+    val proverB = new ErgoLikeTestProvingInterpreter
+    val verifier: ContextEnrichingTestProvingInterpreter = new ContextEnrichingTestProvingInterpreter
+
+    val pubkeyAlice = proverA.dlogSecrets.head.publicImage
+    val pubkeyBob = proverB.dlogSecrets.head.publicImage
+
+    val env = Map("pubkeyA" -> pubkeyAlice, "pubkeyB" -> pubkeyBob)
+    val prop: Values.Value[SSigmaProp.type] = compile(env, """pubkeyA && pubkeyB""").asSigmaProp
+
+    val (rBob, aBob) = DLogInteractiveProver.firstMessage()
+
+    val dlBKnown: Hint = RealCommitment(pubkeyBob, aBob)
+    val bag = HintsBag(Seq(dlBKnown))
+
+    val proofAlice = proverA.prove(prop, ctx, fakeMessage, bag).get
+
+    val bagB = proverB.bagForMultisig(ctx, prop, proofAlice.proof, Seq(pubkeyAlice))
+      .addHint(OwnCommitment(pubkeyBob, rBob, aBob))
+
+    val proofBob = proverB.prove(prop, ctx, fakeMessage, bagB).get
+
+    // Proof generated by Alice without getting Bob's part is not correct
+    verifier.verify(prop, ctx, proofAlice, fakeMessage).get._1 shouldBe false
+
+    // Compound proof from Bob is correct
+    verifier.verify(prop, ctx, proofBob, fakeMessage).get._1 shouldBe true
+  }
+
+  /**
+    * 3-out-of-3 AND signature
+    */
+  property("distributed AND (3 out of 3)") {
+    val proverA = new ErgoLikeTestProvingInterpreter
+    val proverB = new ErgoLikeTestProvingInterpreter
+    val proverC = new ErgoLikeTestProvingInterpreter
+    val verifier: ContextEnrichingTestProvingInterpreter = new ContextEnrichingTestProvingInterpreter
+
+    val pubkeyAlice = proverA.dlogSecrets.head.publicImage
+    val pubkeyBob = proverB.dlogSecrets.head.publicImage
+    val pubkeyCarol = proverC.dlogSecrets.head.publicImage
+
+    val env = Map("pubkeyA" -> pubkeyAlice, "pubkeyB" -> pubkeyBob, "pubkeyC" -> pubkeyCarol)
+    val prop: Values.Value[SSigmaProp.type] = compile(env, """pubkeyA && pubkeyB && pubkeyC""").asSigmaProp
+
+    val (rBob, aBob) = DLogInteractiveProver.firstMessage()
+    val (rCarol, aCarol) = DLogInteractiveProver.firstMessage()
+
+    val dlBKnown: Hint = RealCommitment(pubkeyBob, aBob)
+    val dlCKnown: Hint = RealCommitment(pubkeyCarol, aCarol)
+    val bag = HintsBag(Seq(dlBKnown, dlCKnown))
```

Comment:
Same comment as earlier. `bagA` instead of `bag` is more intuitive, unless there is a specific reason to use this name.

Reply:
done